File: app/src/main/java/com/forkmang/activity/PaymentScreenActivity.java
package com.forkmang.activity;

import static com.forkmang.helper.Constant.SUCCESS_CODE;
import static com.forkmang.helper.Constant.TOKEN_LOGIN;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RelativeLayout;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.forkmang.R;
import com.forkmang.data.RestoData;
import com.forkmang.helper.Constant;
import com.forkmang.helper.StorePrefrence;
import com.forkmang.helper.Utils;
import com.forkmang.models.TableList;
import com.forkmang.network_call.Api;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.json.JSONObject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class PaymentScreenActivity extends AppCompatActivity {
    TableList tableList_get;
    RestoData RestroData;
    String totalpay;
    Context ctx = PaymentScreenActivity.this;
    StorePrefrence storePrefrence;
    String order_id,booking_id, payment_type,isbooktable,order_id_get;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_payment_screen);
        storePrefrence = new StorePrefrence(ctx);
        RelativeLayout relative_view_1 = findViewById(R.id.relative_view_1);
        RelativeLayout relative_view_2 = findViewById(R.id.relative_view_2);
        RadioButton radioButton_cash = findViewById(R.id.radioButton1);
        RadioButton radioButton_online = findViewById(R.id.radioButton2);
        Button btn_payment = findViewById(R.id.btn_payment);

        tableList_get = (TableList) getIntent().getSerializableExtra("model");
        RestroData = (RestoData) getIntent().getSerializableExtra("restromodel");
        totalpay = getIntent().getStringExtra("totalpay");
        isbooktable= getIntent().getStringExtra("isbooktable");

        if(isbooktable.equalsIgnoreCase("yes"))
        {
            booking_id= getIntent().getStringExtra("bookingid");
        }
        else if(isbooktable.equalsIgnoreCase("no")){
            order_id = getIntent().getStringExtra("orderid");
        }

        relative_view_1.setOnClickListener(v -> {
            relative_view_1.setBackgroundColor(ContextCompat.getColor(this, R.color.orange_2));
            relative_view_2.setBackgroundColor(ContextCompat.getColor(this, R.color.white));
            radioButton_cash.setTextColor(ContextCompat.getColor(this, R.color.white));
            radioButton_online.setTextColor(ContextCompat.getColor(this, R.color.black));

            radioButton_online.setChecked(false);
            radioButton_cash.setChecked(true);
        });

        relative_view_2.setOnClickListener(v -> {
            relative_view_2.setBackgroundColor(ContextCompat.getColor(this, R.color.orange_2));
            relative_view_1.setBackgroundColor(ContextCompat.getColor(this, R.color.white));
            radioButton_online.setTextColor(ContextCompat.getColor(this, R.color.white));
            radioButton_cash.setTextColor(ContextCompat.getColor(this, R.color.black));


            radioButton_cash.setChecked(false);
            radioButton_online.setChecked(true);
        });


        btn_payment.setText("Pay - "+ totalpay);

        btn_payment.setOnClickListener(v -> {

            if(radioButton_cash.isChecked())
            {
                payment_type="cash";
            }
            else if(radioButton_online.isChecked())
            {
                payment_type="online";
            }

            if (Utils.isNetworkAvailable(ctx)) {
                if(isbooktable.equalsIgnoreCase("yes"))
                {
                    //callApi_makepayment_1(order_id, payment_type);
                    callApi_makepayment("",booking_id, payment_type,"table");
                }
                else{
                    callApi_makepayment(order_id,storePrefrence.getString(Constant.BOOKINGID), payment_type,"order");
                }



            }
            else{
                Toast.makeText(ctx, Constant.NETWORKEROORMSG, Toast.LENGTH_SHORT).show();
            }
        });

    }


    public void callApi_makepayment(String order_id, String booking_id, String payment_type, String order_type)
    {
        Api.getInfo().make_payment("Bearer "+storePrefrence.getString(TOKEN_LOGIN),"",booking_id, payment_type,"table").
                enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        try{
                            //Log.d("Result", jsonObject.toString());
                            if(response.code() == Constant.SUCCESS_CODE_n)
                            {
                                JSONObject jsonObject = new JSONObject(new Gson().toJson(response.body()));
                                if(jsonObject.getString("status").equalsIgnoreCase(SUCCESS_CODE))
                                {
                                    final Intent mainIntent = new Intent(PaymentScreenActivity.this, Order_ConformationActivity.class);
                                    mainIntent.putExtra("model",tableList_get);
                                    mainIntent.putExtra("restromodel", RestroData);
                                    mainIntent.putExtra("totalpay",totalpay);
                                    if(isbooktable.equalsIgnoreCase("yes"))
                                    {
                                        mainIntent.putExtra("orderid",booking_id);
                                    }
                                    else if(isbooktable.equalsIgnoreCase("no")){
                                        mainIntent.putExtra("orderid",order_id);
                                    }

                                    startActivity(mainIntent);
                                    finish();
                                }
                                else{
                                    Toast.makeText(ctx,jsonObject.getString("message"),Toast.LENGTH_SHORT).show();
                                }

                            }
                            else if(response.code() == Constant.ERROR_CODE)
                            {
                                JSONObject jsonObject = new JSONObject(response.errorBody().string());
                                Toast.makeText(ctx,jsonObject.getString("message"),Toast.LENGTH_SHORT).show();

                            }
                        }
                        catch (Exception ex)
                        {
                            ex.printStackTrace();

                            Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        }
                    }
                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {
                        Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();


                    }
                });
    }

    /*public void callApi_makepayment(String order_id,String payment_type)
    {
        Api.getInfo().make_payment("Bearer "+storePrefrence.getString(TOKEN_LOGIN),order_id, payment_type).
                enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        try{
                            //Log.d("Result", jsonObject.toString());
                            if(response.code() == Constant.SUCCESS_CODE_n)
                            {
                                JSONObject jsonObject = new JSONObject(new Gson().toJson(response.body()));
                                if(jsonObject.getString("status").equalsIgnoreCase(SUCCESS_CODE))
                                {
                                    final Intent mainIntent = new Intent(PaymentScreenActivity.this, Order_ConformationActivity.class);
                                    mainIntent.putExtra("model",tableList_get);
                                    mainIntent.putExtra("restromodel", RestroData);
                                    mainIntent.putExtra("totalpay",totalpay);

                                    mainIntent.putExtra("orderid",order_id);
                                    startActivity(mainIntent);
                                }
                                else{
                                    Toast.makeText(ctx,jsonObject.getString("message"),Toast.LENGTH_SHORT).show();
                                }

                            }
                            else if(response.code() == Constant.ERROR_CODE)
                            {
                                JSONObject jsonObject = new JSONObject(response.errorBody().string());
                                Toast.makeText(ctx,jsonObject.getString("message"),Toast.LENGTH_SHORT).show();

                            }
                        }
                        catch (Exception ex)
                        {
                            ex.printStackTrace();

                            Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();
                        }
                    }
                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {
                        Toast.makeText(ctx, "Error occur please try again", Toast.LENGTH_LONG).show();


                    }
                });
    }*/




}

File: app/src/main/java/com/forkmang/activity/PaymentScreenActivity.kt
package com.forkmang.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.forkmang.R
import com.forkmang.data.RestoData
import com.forkmang.databinding.ActivityPaymentScreenBinding
import com.forkmang.helper.Constant
import com.forkmang.helper.StorePreference
import com.forkmang.helper.Utils
import com.forkmang.helper.showToastMessage
import com.forkmang.models.TableList
import com.forkmang.network_call.Api
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.moyasar.android.sdk.PaymentConfig
import com.moyasar.android.sdk.PaymentResult
import com.moyasar.android.sdk.PaymentSheet
import com.moyasar.android.sdk.payment.models.Payment
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PaymentScreenActivity : AppCompatActivity() {

    var ctx: Context = this@PaymentScreenActivity
    var tableListGet: TableList? = null
    var restroData: RestoData? = null
    var totalpay: Int? = null
    var orderId: String? = null
    var bookingId: String? = null
    var paymentType: String? = null
    var isbooktable: String? = null
    var comingFrom: String? = null
    lateinit var paymentId: String
    lateinit var paymentSheet: PaymentSheet

    private val storePreference by lazy { StorePreference(this) }
    private val binding by lazy { ActivityPaymentScreenBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        comingFrom = intent.getStringExtra("comingfrom")
        if (comingFrom.equals("SelectFood", ignoreCase = true)) {
            tableListGet = intent.getSerializableExtra("model") as TableList?
        } else if (comingFrom.equals("PickupFood", ignoreCase = true)) {
            // not to get table object
        }
        restroData = intent.getSerializableExtra("restromodel") as RestoData?
        totalpay = intent.getIntExtra("totalpay", 0)
        isbooktable = intent.getStringExtra("isbooktable")
        if (isbooktable.equals("yes", ignoreCase = true)) {
            bookingId = intent.getStringExtra("bookingid")
        } else if (isbooktable.equals("no", ignoreCase = true)) {
            orderId = intent.getStringExtra("orderid")
        }
        val config = PaymentConfig(
            amount = totalpay ?: 0,
            currency = "SAR",
            description = "Sample Android SDK Payment",
            apiKey = "pk_test_H3rN5E6jNwTiYunxabgvUSgRUraz8MaQn6ZSVNdj"
        )

        paymentSheet = PaymentSheet(this, { handlePaymentResult(it) }, config)

        binding.relativeView1.setOnClickListener {
            binding.relativeView1.setBackgroundColor(ContextCompat.getColor(this, R.color.orange_2))
            binding.relativeView2.setBackgroundColor(ContextCompat.getColor(this, R.color.white))
            binding.radioButton1.setTextColor(ContextCompat.getColor(this, R.color.white))
            binding.radioButton2.setTextColor(ContextCompat.getColor(this, R.color.black))
            binding.radioButton2.isChecked = false
            binding.radioButton1.isChecked = true
        }
        binding.relativeView2.setOnClickListener {
            binding.relativeView2.setBackgroundColor(ContextCompat.getColor(this, R.color.orange_2))
            binding.relativeView1.setBackgroundColor(ContextCompat.getColor(this, R.color.white))
            binding.radioButton2.setTextColor(ContextCompat.getColor(this, R.color.white))
            binding.radioButton1.setTextColor(ContextCompat.getColor(this, R.color.black))
            binding.radioButton1.isChecked = false
            binding.radioButton2.isChecked = true
        }
        binding.btnPayment.text = "Pay - $totalpay"
        binding.btnPayment.setOnClickListener {
            paymentType = if (binding.radioButton1.isChecked) {
                "cash"
            } else if (binding.radioButton2.isChecked) {
                "online"
            } else {
                showToastMessage("Please select a mode of payment")
                return@setOnClickListener
            }
            if (Utils.isNetworkAvailable(ctx)) {
                if (isbooktable.equals("yes", ignoreCase = true)) {
                    callApiMakePayment("", bookingId, paymentType, "table", totalpay.toString())
                } else {
                    if (comingFrom.equals("SelectFood", ignoreCase = true)) {
                        callApiMakePayment(
                            orderId,
                            storePreference.getString(Constant.BOOKINGID),
                            paymentType,
                            "order",
                            totalpay.toString()
                        )
                    } else if (comingFrom.equals("PickupFood", ignoreCase = true)) {
                        // not to get table object
                        callApiMakePayment(orderId, "", paymentType, "order", totalpay.toString())
                    }
                }

            } else {
                showToastMessage(Constant.NETWORKEROORMSG)
            }
        }
    }

    private fun handlePaymentResult(result: PaymentResult) {
        when (result) {
            is PaymentResult.Completed -> {
                handleCompletedPayment(result.payment);
            }
            is PaymentResult.Failed -> {
                showToastMessage(result.error.message ?: "")
            }
            PaymentResult.Canceled -> {
                // User has canceled the payment
                showToastMessage("Payment canceled")
            }
        }
    }

    private fun handleCompletedPayment(payment: Payment) {
        when (payment.status) {
            "paid" -> {
                callGetPaymentDetails(
                    payment.id,
                    payment.status,
                    payment.amount,
                    payment.currency,
                    payment.invoiceId,
                    payment.source
                )

            }
            "failed" -> {
                val errorMessage = payment.source["message"]
                showToastMessage(errorMessage.toString())
            }
            else -> { /* Handle other statuses */
            }
        }
    }

    private fun callGetPaymentDetails(
        id: String,
        status: String,
        amount: Int,
        currency: String,
        invoiceId: String?,
        source: MutableMap<String, String>
    ) {
        Api.info.getPaymentData(
            "Bearer " + storePreference.getString(Constant.TOKEN_LOGIN),
            id, status, amount.toString(), currency, invoiceId ?: "", source, paymentId

        )?.enqueue(object : Callback<JsonObject?> {
            override fun onResponse(call: Call<JsonObject?>, response: Response<JsonObject?>) {
                try {
                    if (response.code() == Constant.SUCCESS_CODE_n) {
                        val jsonObject = JSONObject(Gson().toJson(response.body()))
                        if (jsonObject.getString("status")
                                .equals(Constant.SUCCESS_CODE, ignoreCase = true)
                        ) {
                            val mainIntent = Intent(
                                this@PaymentScreenActivity,
                                OrderConformationActivity::class.java
                            )
                            if (comingFrom.equals("SelectFood", ignoreCase = true)) {
                                mainIntent.putExtra("model", tableListGet)
                                mainIntent.putExtra("comingfrom", "SelectFood")
                            } else if (comingFrom.equals("PickupFood", ignoreCase = true)) {
                                mainIntent.putExtra("comingfrom", "PickupFood")
                            }
                            mainIntent.putExtra("restromodel", restroData)
                            mainIntent.putExtra("totalpay", totalpay)
                            if (isbooktable.equals("yes", ignoreCase = true)) {
                                mainIntent.putExtra("orderid", bookingId)
                            } else if (isbooktable.equals("no", ignoreCase = true)) {
                                mainIntent.putExtra("orderid", orderId)
                            }
                            startActivity(mainIntent)
                            finish()
                        } else {
                            showToastMessage(jsonObject.getString("message"))
                        }
                    } else if (response.code() == Constant.ERROR_CODE) {
                        val jsonObject = JSONObject(response.errorBody()?.string())
                        showToastMessage(jsonObject.getString("message"))
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    showToastMessage(Constant.ERRORMSG)
                }
            }
            override fun onFailure(call: Call<JsonObject?>, t: Throwable) {
                showToastMessage(Constant.ERRORMSG)
            }

        })
    }

    private fun callApiMakePayment(
        order_id: String?,
        booking_id: String?,
        payment_type: String?,
        order_type: String?,
        amt: String
    ) {
        Api.info.makePayment(
            "Bearer " + storePreference.getString(Constant.TOKEN_LOGIN),
            order_id,
            booking_id,
            payment_type,
            order_type,
            amt
        )?.enqueue(object : Callback<JsonObject?> {
            override fun onResponse(call: Call<JsonObject?>, response: Response<JsonObject?>) {
                try {
                    if (response.code() == Constant.SUCCESS_CODE_n) {
                        val jsonObject = JSONObject(Gson().toJson(response.body()))
                        if (jsonObject.getString("status")
                                .equals(Constant.SUCCESS_CODE, ignoreCase = true)
                        ) {
                            paymentId = jsonObject.getJSONObject("data").getString("payment_id")
                            paymentSheet.present()
                        } else {
                            showToastMessage(jsonObject.getString("message"))
                        }
                    } else if (response.code() == Constant.ERROR_CODE) {
                        val jsonObject = JSONObject(response.errorBody()?.string())
                        showToastMessage(jsonObject.getString("message"))
                    }
                } catch (ex: Exception) {
                    ex.printStackTrace()
                    showToastMessage(Constant.ERRORMSG)
                }
            }

            override fun onFailure(call: Call<JsonObject?>, t: Throwable) {
                showToastMessage(Constant.ERRORMSG)
            }
        })
    }

}